crypto_secretstream/src/push.rs

```diff
@@ -0,0 +1,96 @@
+use chacha20poly1305::{
+    aead::{generic_array::GenericArray, AeadInPlace, Buffer, NewAead},
+    XChaCha20Poly1305,
+};
+use rand_core::{CryptoRng, RngCore};
+use snafu::OptionExt;
+
+use super::nonce::Nonce;
+use crate::Key;
+
+#[derive(Debug, snafu::Snafu)]
+pub enum Error {
+    #[snafu(display("encryption failed"))]
+    Encrypt,
+}
+
+enum State {
+    Init {
+        nonce: Nonce,
+        key: Key,
+    },
+    Running {
+        nonce: Nonce,
+        aead: XChaCha20Poly1305,
+    },
+}
+
+/// Stream that can encrypt messages to be decrypted by [`crate::PullStream`]
+pub struct Stream(State);
+
+impl Stream {
+    /// Create new stream for sending messages with a preshared key
+    ///
+    /// The RNG is needed to generate a nonce
+    pub fn init<T>(csprng: &mut T, key: Key) -> Self
+    where
+        T: RngCore + CryptoRng,
+    {
+        Self(State::Init {
+            nonce: Nonce::generate(csprng),
+            key,
+        })
+    }
+
+    /// Encrypt a buffer filled with a message, and its associated data
```

Comment:
```suggestion
    /// Encrypt a buffer filled with a message and its associated data.
```

Also mention that the buffer will be overwritten with the encrypted data.